proparse/src/main/java/org/prorefactor/proparse/antlr4/Lexer.java

```diff
@@ -0,0 +1,1117 @@
+
+  private static final int EOF_CHAR = -1;
+
+  /** Lowercase value of current character */
+  private int currChar;
+
+  /** Current character, before being lowercased */
+  private int currInt;
+  private int currFile, currLine, currCol;
+  private int prevFile, prevLine, prevCol;
+  
+  private int currStringType;
+  private StringBuilder currText = new StringBuilder();
+
+  private ProgressLexer prepro;
+
+  private boolean gettingAmpIfDefArg = false;
+  private boolean preserve = false;
+  private int preserveFile;
+  private int preserveLine;
+  private int preserveCol;
+  private int preserveSource;
+  private int preserveChar;
+
+  private int textStartFile;
+  private int textStartLine;
+  private int textStartCol;
+  private int textStartSource;
+
+  private Set<Integer> comments = new HashSet<>();
+  private Set<Integer> loc = new HashSet<>();
+
+  Lexer(ProgressLexer prepro) throws IOException {
+    this.prepro = prepro;
+    getChar(); // We always assume "currChar" is available.
+  }
+
+
+  //////////////// Lexical productions listed first, support functions follow.
+  public ProToken nextToken() {
+    LOGGER.trace("Entering nextToken()");
+    try {
+    for (;;) {
+
+      if (preserve) {
+        // The preserved character is the character prior to currChar.
+        textStartFile = preserveFile;
+        textStartLine = preserveLine;
+        textStartCol = preserveCol;
+        textStartSource = preserveSource;
+        currText.setLength(1);
+        currText.setCharAt(0, (char) preserveChar);
+        preserveDrop(); // we are done with the preservation
+        switch (preserveChar) {
+          case '.':
+            return periodStart();
+          case ':':
+            return colon();
+        } // switch
+      }
+
+      // Proparse Directive
+      // Check this before setting currText...
+      // we don't want BEGIN_PROPARSE_DIRECTIVE in the text
+      if (currInt == ProgressLexer.PROPARSE_DIRECTIVE) {
+        textStartFile = prepro.getTextStart().getFile();
+        textStartLine = prepro.getTextStart().getLine();
+        textStartCol = prepro.getTextStart().getCol();
+        textStartSource = prepro.getTextStart().getSourceNum();
+        getChar();
+        return makeToken(PreprocessorParser.PROPARSEDIRECTIVE, prepro.getProparseDirectiveText());
+      }
+
+      textStartFile = prepro.getFileIndex();
+      textStartLine = prepro.getLine2();
+      textStartCol = prepro.getColumn();
+      textStartSource = prepro.getSourceNum();
+      currText.setLength(1);
+      currText.setCharAt(0, (char) currInt);
+
+      if (gettingAmpIfDefArg) {
+        getChar();
+        gettingAmpIfDefArg = false;
+        return ampIfDefArg();
+      }
+
+      switch (currChar) {
+
+        case '\t':
+        case '\n':
+        case '\f':
+        case '\r':
+        case ' ':
+          getChar();
+          return whitespace();
+
+        case '"':
+        case '\'':
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1151)